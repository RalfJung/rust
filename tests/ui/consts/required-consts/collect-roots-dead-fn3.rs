//@revisions: noopt opt
//@ build-fail
//@[noopt] compile-flags: -Copt-level=0
//@[opt] compile-flags: -O

#![crate_type = "lib"]
#![feature(inline_const)]

// Will be `inline` with optimizations, so then `g::<()>` becomes reachable. At the same time `g` is
// not "mentioned" in `f` since it is only called inside an inline `const` and hence never appears
// in its MIR! This fundamentally is an issue caused by reachability walking the HIR (before inline
// `const` are extracted) and collection being based on MIR.
pub fn f() {
    loop {}; const { g::<()>() }
}

// When this comes reachable (for any `T`), so does `hidden_root`.
// And `hidden_root` is monomorphic so it can become a root!
const fn g<T>() {
    match std::mem::size_of::<T>() {
        0 => (),
        _ => hidden_root(),
    }
}

#[inline(never)]
const fn hidden_root() {
    fail::<()>();
}

#[inline(never)]
const fn fail<T>() {
    // Hiding this in a generic fn so that it doesn't get evaluated by
    // MIR passes.
    const { panic!(); } //~ERROR: evaluation of `fail::<()>::{constant#0}` failed
}
