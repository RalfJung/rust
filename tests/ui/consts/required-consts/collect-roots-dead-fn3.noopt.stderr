error[E0080]: evaluation of `fail::<()>::{constant#0}` failed
  --> $DIR/collect-roots-dead-fn3.rs:35:13
   |
LL |     const { panic!(); }
   |             ^^^^^^^^ the evaluated program panicked at 'explicit panic', $DIR/collect-roots-dead-fn3.rs:35:13
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
  --> $DIR/collect-roots-dead-fn3.rs:35:5
   |
LL |     const { panic!(); }
   |     ^^^^^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn fail::<()>`
  --> $DIR/collect-roots-dead-fn3.rs:28:5
   |
LL |     fail::<()>();
   |     ^^^^^^^^^^^^

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
